src/vaults.ts

```diff
@@ -1,59 +1,225 @@
+  await Promise.all(
+    Array.from(vaults.values()).map(async ({ oracleType, collateralPriceFeed }) => {
+      if (oracleType != 0)
+        throw new Error('Oracle not supported on on this version, please update liquidator')
+
+      let price = 1
+      if (!collateralPriceFeed.equals(DEFAULT_PUBLIC_KEY)) {
+        const account = await connection.getAccountInfo(collateralPriceFeed)
+
+        const { price: fetchedPrice } = parsePriceData(account.data)
+
+        price = fetchedPrice
+      }
+
+      if (collateralPrices.has(collateralPriceFeed.toString())) return
+
+      // checking again here because it could have changed while fetching
+      if (collateralPrices.has(collateralPriceFeed.toString())) return
+      collateralPrices.set(collateralPriceFeed.toString(), new BN(price * 10 ** ORACLE_OFFSET))
+    })
+  )
+
+  console.log('Calculating..')
+  let i = 0
+
+  for (const entry of entries) {
+    if (!vaults.has(entry.vault.toString())) continue
+
+    const vault = vaults.get(entry.vault.toString())
+    const collateralPrice = collateralPrices.get(vault.collateralPriceFeed.toString())
+    const syntheticPrice = prices.getPriceFor(vault.synthetic).val
+    const amount = getAmountForLiquidation(entry, vault, collateralPrice, syntheticPrice)
+
+    if (amount.val.eqn(0)) continue
+
+    console.log(amount.val.toString())
+
+    console.log('Found account for liquidation')
+    const syntheticToken = new Token(connection, vault.synthetic, TOKEN_PROGRAM_ID, wallet)
+    const collateralToken = new Token(connection, vault.collateral, TOKEN_PROGRAM_ID, wallet)
+
+    const [xUSDAccount, syntheticAccount, collateralAccount] = await Promise.all([
+      xUSDToken.getOrCreateAssociatedAccountInfo(wallet.publicKey),
+      syntheticToken.getOrCreateAssociatedAccountInfo(wallet.publicKey),
+      collateralToken.getOrCreateAssociatedAccountInfo(wallet.publicKey)
+    ])
+
+    const isUsdTheSynthetic = xUSDAccount.address.equals(syntheticAccount.address)
+
+    console.log('Preparing synthetic for liquidation..')
+
+    // needed value + 2% to account for swap fee and price fluctuations
+    const value = amountToValue(amount, syntheticPrice)
+
+    const neededAmount = toDecimal(
+      amount.val.muln(102).divn(100).sub(syntheticAccount.amount),
+      amount.scale
+    )
+
+    // Minimum amount that can be traded on synthetify
+    const swapAmount = neededAmount.val.gt(new BN(1000)) ? neededAmount.val : new BN(1001)
+
+    if (swapAmount.gt(xUSDAccount.amount)) throw new Error('not enough xUSD')
+
+    let tx = new Transaction().add(await exchange.updatePricesInstruction(state.account.assetsList))
+
+    // Swap to the right synthetic
+    if (!isUsdTheSynthetic && neededAmount.val.gten(0)) {
+      console.log('Swapping synthetics..')
+
+      tx.add(
+        Token.createApproveInstruction(
+          TOKEN_PROGRAM_ID,
+          xUSDAccount.address,
+          exchange.exchangeAuthority,
+          wallet.publicKey,
```

Comment:
What if our account on token does not exist ?

Reply:
on line 143 all of them are created